Ladybird/AppKit/UI/LadybirdWebViewBridge.cpp

```diff
@@ -0,0 +1,323 @@
+{
+    client().async_mouse_move(to_content_position(position), 0, to_underlying(button), modifiers);
+}
+
+void WebViewBridge::mouse_double_click_event(Gfx::IntPoint position, GUI::MouseButton button, KeyModifier modifiers)
+{
+    client().async_doubleclick(to_content_position(position), button, to_underlying(button), modifiers);
+}
+
+void WebViewBridge::key_down_event(KeyCode key_code, KeyModifier modifiers, u32 code_point)
+{
+    client().async_key_down(key_code, modifiers, code_point);
+}
+
+void WebViewBridge::key_up_event(KeyCode key_code, KeyModifier modifiers, u32 code_point)
+{
+    client().async_key_up(key_code, modifiers, code_point);
+}
+
+Optional<WebViewBridge::Paintable> WebViewBridge::paintable()
+{
+    Gfx::Bitmap* bitmap = nullptr;
+    Gfx::IntSize bitmap_size;
+
+    if (m_client_state.has_usable_bitmap) {
+        bitmap = m_client_state.front_bitmap.bitmap.ptr();
+        bitmap_size = m_client_state.front_bitmap.last_painted_size;
+    } else {
+        bitmap = m_backup_bitmap.ptr();
+        bitmap_size = m_backup_bitmap_size;
+    }
+
+    if (!bitmap)
+        return {};
+    return Paintable { *bitmap, bitmap_size };
+}
+
+void WebViewBridge::notify_server_did_layout(Badge<WebView::WebContentClient>, Gfx::IntSize content_size)
+{
+    if (on_layout) {
+        content_size = scale_for_device(content_size, m_inverse_device_pixel_ratio);
+        on_layout(content_size);
+    }
+}
+
+void WebViewBridge::notify_server_did_paint(Badge<WebView::WebContentClient>, i32 bitmap_id, Gfx::IntSize size)
+{
+    if (m_client_state.back_bitmap.id == bitmap_id) {
+        m_client_state.has_usable_bitmap = true;
+        m_client_state.back_bitmap.pending_paints--;
+        m_client_state.back_bitmap.last_painted_size = size;
+        swap(m_client_state.back_bitmap, m_client_state.front_bitmap);
+        // We don't need the backup bitmap anymore, so drop it.
+        m_backup_bitmap = nullptr;
+
+        if (on_ready_to_paint)
+            on_ready_to_paint();
+
+        if (m_client_state.got_repaint_requests_while_painting) {
+            m_client_state.got_repaint_requests_while_painting = false;
+            request_repaint();
+        }
+    }
+}
+
+void WebViewBridge::notify_server_did_invalidate_content_rect(Badge<WebView::WebContentClient>, Gfx::IntRect const&)
+{
+    request_repaint();
+}
+
+void WebViewBridge::notify_server_did_change_selection(Badge<WebView::WebContentClient>)
+{
+    request_repaint();
+}
+
+void WebViewBridge::notify_server_did_request_cursor_change(Badge<WebView::WebContentClient>, Gfx::StandardCursor cursor)
+{
+    if (on_cursor_change)
+        on_cursor_change(cursor);
+}
+
+void WebViewBridge::notify_server_did_request_scroll(Badge<WebView::WebContentClient>, i32 x_delta, i32 y_delta)
+{
+    auto position = m_viewport_rect.location();
+    position.set_x(position.x() + x_delta);
+    position.set_y(position.y() + y_delta);
+
```

Comment:
I would verify that the scroll logic in the WebView is consistent with the changes I merged earlier today to how this works (esp the need(?) to add the scroll increment to some places?

Reply:
We don't actually propagate mouse wheel events at all right now, so this isn't actually hit. I'll add a fixme to make sure it's right when we do propagate wheel events.